ragna/deploy/_cli/core.py

```diff
@@ -171,3 +177,215 @@ def wait_for_api() -> None:
+        document_filesystem_handler_class = DocumentFilesystemHandlerImport(
+            document_filesystem_handler_class=document_filesystem_handler
+        ).document_filesystem_handler_class
+    except ValidationError:
+        raise typer.BadParameter(
+            f"Unable to import document filesystem handler: {document_filesystem_handler}"
+        )
+
+    try:
+        getattr(document_filesystem_handler_class, "from_path")
+    except AttributeError:
+        raise typer.BadParameter(
+            f"Document filesystem handler {document_filesystem_handler} does not support ingesting documents. Please implement a `from_path` method."
+        )
+
+    make_session = database.get_sessionmaker(database_url)
+
+    BATCH_SIZE = 10
+    bad_documents = []
+    with tqdm(total=len(documents) // BATCH_SIZE) as pbar:
+        for batch_number in range(0, len(documents), BATCH_SIZE):
+            document_collection = []
+            document_metadata_collection = []
+
+            for document in documents[batch_number : batch_number + BATCH_SIZE]:
+                try:
+                    doc_instance = document_filesystem_handler_class.from_path(document)
+                    document_collection.append(doc_instance)
+                    document_metadata_collection.append(
+                        (
+                            schemas.Document.from_core(doc_instance),
+                            doc_instance.metadata,
+                        )
+                    )
+                except Exception:
+                    bad_documents.append(document)
+
+            if not document_collection:
+                continue
+
+            # TODO: add a check to see if the documents already exist in the database
+            # and skip them if they do - this will avoid the expensive operation of
+            # computing embeddings and storing them in the source storage
+
+            try:
+                database.add_documents(
+                    make_session(),
+                    user=default_user(),
+                    document_metadata_collection=document_metadata_collection,
+                )
+            except Exception:
+                bad_documents.extend(
+                    documents[batch_number : batch_number + BATCH_SIZE]
+                )
+
+            try:
+                source_storage_class().store(corpus_name, document_collection)
+            except Exception:
+                # TODO: should rollback the ragna database transaction here as well
+                # to ensure that the database and source storage are in sync
+                bad_documents.extend(
+                    documents[batch_number : batch_number + BATCH_SIZE]
+                )
+
+            pbar.update(1)
+
+        if verbose:
+            typer.echo(
+                f"Failed to ingest the following documents: {set(bad_documents)}"
+            )
+
+
+@corpus_app.command(help="List all available corpora.", name="list")
+def list_(
```

Comment:
`list`, `delete` and `metadata` would probably require extending the interface/protocol methods in some way, so I didn't explore this any further.